sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffleCoalesceExec.scala

```diff
@@ -423,12 +516,55 @@ abstract class CoalesceIteratorBase[T <: AutoCloseable : ClassTag, R](
   override def close(): Unit = {
     serializedTables.forEach(_.close())
     serializedTables.clear()
+    pendingResultIter.foreach(_.foreach {
+      case ac: AutoCloseable => ac.safeClose()
+      case _ => // non-closeable result
+    })
+    pendingResultIter = None
+  }
+
+  /**
+   * Get the next result, checking for pending results from splits first.
+   * Returns Some(result) if there's a result available, None if we need
+   * to concatenate a new batch.
+   */
+  protected def getNextResult(): Option[R] = {
+    pendingResultIter.flatMap { iter =>
+      if (iter.hasNext) {
+        Some(iter.next())
+      } else {
+        pendingResultIter = None
+        None
+      }
+    }
   }
 
   protected def concatenateTables(tables: ArrayBuffer[T]): R = {
     withResource(new MetricRange(concatTimeMetric)) { _ =>
-      withRetryNoSplit(tables.toSeq) { tablesSeq =>
-        tableOperator.concat(tablesSeq.toArray)
+      val tablesSeq = tables.toSeq
+      splitPolicy match {
+        case Some(policy) =>
+          // Create AutoCloseableTargetSize with targetBatchByteSize and minSplitSize
+          val targetSizeWrapper = AutoCloseableTargetSize(targetBatchByteSize, minSplitSizeForRetry)
+          val wrapper = CloseableTableSeqWithTargetSize(tablesSeq, targetSizeWrapper)
+          val resultIter = withRetry(wrapper, policy) { wrappedSeq =>
+            tableOperator.concat(wrappedSeq.toArray)
+          }
+          // Store the iterator so we can yield results one by one if splits occurred
+          if (resultIter.hasNext) {
+            val firstResult = resultIter.next()
+            // If there are more results, store the iterator for subsequent calls
+            if (resultIter.hasNext) {
+              pendingResultIter = Some(resultIter)
+            }
+            firstResult
+          } else {
+            throw new IllegalStateException("withRetry returned empty iterator")
```

Comment:
we should closeOnExcept firstResult then, and make sure `CoalesceIteratorBase` is closed in this case.

Reply:
Yes, I pushed that change